### ISSUE TITLE
mkDerivation version from VERSION file appends a dash (-) to derivation path

### ISSUE THREAD

**Issue by prtzl** [NixOS/nixpkgs#258969]
### Describe the bug
Setting the value of `version` field of `mkDerivation` by reading from a file using `builtins.readFile` results with the derivation path appended with `-`.

### Steps To Reproduce
Create a short derivation:
```nix
der = pkgs.stdenv.mkDerivation rec {
  pname = "der";
  # version = "v1.0";
  version = builtins.readFile ./VERSION;

  src = ./.;

  installPhase = ''
    echo -n "${pname}-${version}" >> $out
  '';
};
```

And a VERSION file:
```
v1.0
```
Run `nix build && readlink result`

I have tested with following permutations of the version string for both the file and the string literal:
* v1.0
* v1.0.0
* 1.0
* 1.0.0
* 1
* v1

### Expected behavior
The path link should be the same if I use the `version = "v1.0"` or the `readFile` method.

### Screenshots
Result file path when running version of string literal:
```
/nix/store/q0bf4laaf3j35znz1cyilf9xgqvr5pv0-der-v1.0
/nix/store/zgbxywn1ads6p9nqrcr54allpb2qj630-der-1.0
/nix/store/sr0hzmgjkqibrp10dlbd8dwbwrzq0v62-der-1.0.0
/nix/store/hf7y1cfix21ls6dya1l2s0a73mpc2q2c-der-v1.0.0
/nix/store/ai1w3n50xnv27rw3pppq3vzq30k1yqs2-der-v1
/nix/store/140nckimbd27ap4byapi8pq56msjfngn-der-1
```

Result file path when running version from file:
```
/nix/store/0hzdfdyjq2vj7xx1x3q5n00h7d2w0zhs-der-v1.0-
/nix/store/sf72iyfqxqim8knkmgdmraxn5g9scba4-der-1.0-
/nix/store/l5lcg4jkv145gz2wlnjw423pg2mi48yi-der-1.0.0-
/nix/store/zvn5v2lc2bzz7yfn1h9s3gxijwj6ad67-der-v1.0.0-
/nix/store/ghqs7sn6wryfxd354nbv8gs281kd1lsg-der-v1-
/nix/store/q9knabik58a66bgiqfj7xbmqs1f6y1yc-der-1-
```

### Metadata

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.5.3, NixOS, 23.05 (Stoat), 23.05.20230917.5d017a8`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.17.0`
 - channels(matej): `"nixpkgs"`
 - nixpkgs: `/home/matej/.nix-defexpr/channels/nixpkgs`
```

**Comment by tu-maurice** [comment]
Is it possible that the file contains a newline that gets replaced with a dash during sanitization?

**Comment by prtzl** [comment]
> Is it possible that the VERSION file contains a newline that gets replaced with a dash during sanitization?

Yes, I did not think of that, since I used `echo -n "v1.0" > VERSION` to create the file. From then I used my editor to modify and test the file.
Using `cat -A VERSION` when creating the file I get:
```console
v1.0%
```
And when just saving it in my editor (nvim) I get:
```console
v1.0$
```
So nvim in my configuration adds line feed (LF) character when saved, so it does not show as an extra line. I guess that is what gets converted to `-`.
